src/btree/row_modify.c

```diff
@@ -324,10 +329,14 @@ __wt_update_obsolete_check(
              */
             if (upd->start_ts != WT_TS_NONE && upd->start_ts >= oldest && upd->start_ts < stable)
                 ++upd_unstable;
-        } else if (first == NULL && upd->type == WT_UPDATE_BIRTHMARK)
-            first = prev;
-        else if (first == NULL && WT_UPDATE_DATA_VALUE(upd))
-            first = upd;
+        } else {
+            if (first == NULL && upd->type == WT_UPDATE_BIRTHMARK)
```

Comment:
The fix looks correct to me. It deserves a comment now, the logic in here is complex, so we should add a comment to describe why.

I'm also not a huge fan of ternary (?) statements, especially when they are nested within other conditional blocks. Could this be restructured? Maybe:

```
if (first == NULL) {
    if (upd_visible_all_seen && upd->type == WT_UPDATE_BIRTHMARK)
        first = prev;
    else if (upd->type == WT_UPDATE_BIRTHMARK || WT_UPDATE_DATA_VALUE(upd))
        first = upd;
}
```

It adds another level of nesting, but I think the flow is easier to follow.